### PR TITLE
ISPN-9625 Configurable clustered locks container

### PR DESCRIPTION
**Pull request by karesti**
https://issues.jboss.org/browse/ISPN-9625

backport : https://github.com/infinispan/infinispan/pull/6607

### CODE REVIEW

File: lock/src/main/java/org/infinispan/lock/configuration/ClusteredLockManagerConfigurationBuilder.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+ * The {@link org.infinispan.lock.api.ClusteredLockManager} configuration builder.
+ * <p>
+ * It configures the number of owner, the {@link Reliability} mode and the {@link Consistency}.
+ *
+ * @author Katia Aresti, karesti@redhat.com
+ * @since 10.0
+ */
+public class ClusteredLockManagerConfigurationBuilder implements Builder<ClusteredLockManagerConfiguration> {
+
+   private static final ClusteredLockManagerConfiguration DEFAULT = new ClusteredLockManagerConfigurationBuilder(null).create();
```
[the remark this answers is not in the document]
Comment (karesti) [reply lock/src/main/java/org/infinispan/lock/configuration/ClusteredLockManagerConfigurationBuilder.java]: Thanks for the review, I'm finishing this PR. This code is inspired from counters

File: lock/src/main/java/org/infinispan/lock/configuration/ClusteredLockManagerConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+ * @since 10.0
+ */
+public class ClusteredLockManagerConfiguration {
+
+   private static final Log log = LogFactory.getLog(ClusteredLockManagerConfiguration.class, Log.class);
+   static final AttributeDefinition<Reliability> RELIABILITY = AttributeDefinition
+         .builder("reliability", Reliability.CONSISTENT)
+         .validator(value -> {
+            if (value == null) {
+               throw log.invalidReliabilityMode();
```
[the remark this answers is not in the document]
Comment (karesti) [reply lock/src/main/java/org/infinispan/lock/configuration/ClusteredLockManagerConfiguration.java]: This is for the parser

**Comment by karesti** [comment]
@danberindei ok to delete de consistency attribute

**Comment by karesti** [comment]
@pruivo @danberindei should be mergeable

**Comment by karesti** [comment]
@danberindei I think I addressed all your comments, when you have some time I would like to merge this as soon as possible 🙏

File: lock/src/main/java/org/infinispan/lock/configuration/ClusteredLockManagerConfigurationBuilder.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (danberindei) [reply lock/src/main/java/org/infinispan/lock/configuration/ClusteredLockManagerConfigurationBuilder.java]: I noticed that counters use the same approach, but you don't actually use the `builder` field.

File: lock/src/main/java/org/infinispan/lock/configuration/Reliability.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package org.infinispan.lock.configuration;
+
+/**
+ * Locks are stored in a container that can privilege availability or consistency.
+ * Availability is useful when you wish to be able to use the lock most of the time.
```
Comment (danberindei) [inline lock/src/main/java/org/infinispan/lock/configuration/Reliability.java]: I'm pretty sure everyone wants to be able to use the lock **at least** most of the time ;)

I would say something like

```
Locks are stored in a container that can privilege availability or consistency.

Most of the time, locks are both available and consistent. 
But in some situations, e.g. when the cluster splits, there is a choice between keeping the locks available everywhere (potentially allowing multiple nodes to acquire the same lock)
or making it unavailable in the minority partition(s) (potentially requiring administrator intervention to become available again).
@see org.infinispan.partitionhandling.PartitionHandling
```

File: lock/src/main/java/org/infinispan/lock/logging/Log.java [hunk 4]
```diff
@@ -26,4 +27,21 @@
...
+   /**
+    * Configuration exceptions
+    *
+    * @since 10.0
+    */
+
+   @Message(value = "Missing name for the clustered lock", id = 29004)
+   ClusteredLockException missingName();
+
+   @Message(value = "Invalid number of owner. It must be higher than zero but it was %s", id = 29005)
```
Comment (danberindei) [inline lock/src/main/java/org/infinispan/lock/logging/Log.java]:
```suggestion
   @Message(value = "Invalid number of owner. It must be higher than zero or -1 but it was %s", id = 29005)
```

File: lock/src/main/resources/schema/infinispan-clustered-locks-config-10.0.xsd [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                    <xs:documentation>
+                        Configures a clustered lock.
+                    </xs:documentation>
+                </xs:annotation>
+            </xs:element>
+        </xs:sequence>
+        <xs:attribute name="num-owners" type="xs:integer" default="-1">
+            <xs:annotation>
+                <xs:documentation>
+                    Sets the number of copies of the clustered lock's status available in the cluster.
```
Comment (danberindei) [inline lock/src/main/resources/schema/infinispan-clustered-locks-config-10.0.xsd]:
```suggestion
                    Sets the number of copies of the clustered lock's status to keep in the cluster, or -1 to keep it on all the nodes.
```
Please also remove the next line

File: lock/src/main/resources/schema/infinispan-clustered-locks-config-10.0.xsd [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                    <xs:documentation>
+                        Only the majority partition is allowed to read and update the clustered lock if it is available on that partition.
+                        The remaining partitions are only allowed to read if the clustered lock is available on that partition.
+                    </xs:documentation>
+                </xs:annotation>
+            </xs:enumeration>
+        </xs:restriction>
+    </xs:simpleType>
+
+    <xs:simpleType name="consistency">
```
Comment (danberindei) [inline lock/src/main/resources/schema/infinispan-clustered-locks-config-10.0.xsd]: Remove

File: lock/src/test/java/org/infinispan/lock/ClusteredLockSplitBrainLowConsistencyTest.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+import org.infinispan.configuration.cache.CacheMode;
+import org.infinispan.lock.api.ClusteredLock;
+import org.infinispan.lock.api.ClusteredLockManager;
+import org.infinispan.lock.configuration.Reliability;
+import org.infinispan.manager.EmbeddedCacheManager;
+import org.testng.annotations.Test;
+
+@Test(groups = "functional", testName = "clusteredLock.ClusteredLockSplitBrainLowConsistencyTest")
+public class ClusteredLockSplitBrainLowConsistencyTest extends BaseClusteredLockSplitBrainTest {
```
Comment (danberindei) [inline lock/src/test/java/org/infinispan/lock/ClusteredLockSplitBrainLowConsistencyTest.java]: Please rename to `BaseClusteredLockSplitBrainTest`. There are several other tests and test configuration files named based on "consistency", I suggest using `Dist` everywhere (and `Repl` for the others).

File: lock/src/main/java/org/infinispan/lock/configuration/Reliability.java (hunk 3 above)
Comment (karesti) [reply lock/src/main/java/org/infinispan/lock/configuration/Reliability.java]: Ok! 😄

File: lock/src/test/java/org/infinispan/lock/ClusteredLockSplitBrainLowConsistencyTest.java (hunk 7 above)
Comment (karesti) [reply lock/src/test/java/org/infinispan/lock/ClusteredLockSplitBrainLowConsistencyTest.java]: I renamed

**Comment by karesti** [comment]
I added the 9.4 xsd and wrote since 9.4, because this needs to be backported

**Comment by karesti** [comment]
@danberindei I reported all your changes and squashed all the commits

File: documentation/src/main/asciidoc/user_guide/clustered_lock.adoc [hunk 8]
```diff
@@ -117,6 +115,7 @@ Releases - or unlocks - a `ClusteredLock`, if such exists, no matter who is hold
 Calling this method may cause concurrency issues and has to be used in *exceptional situations*.
 
 [[clustered_lock_interface]]
+
```
[the remark this answers is not in the document]
Comment (karesti) [reply documentation/src/main/asciidoc/user_guide/clustered_lock.adoc]: I don't think the line is needed, I'm removing

File: lock/src/main/java/org/infinispan/lock/configuration/ClusteredLockManagerConfiguration.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (karesti) [reply lock/src/main/java/org/infinispan/lock/configuration/ClusteredLockManagerConfiguration.java]: In the counters we do it this way too. I'm not doing any additional change in this PR, but in another PR as an improvement if necessary

**Comment by danberindei** [comment]
Integrated, thanks Katia!

**Pull request merged**